numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -517,6 +539,31 @@ PyUFunc_SimpleUniformOperationTypeResolver(
             out_dtypes[0] = ensure_dtype_nbo(PyArray_DESCR(operands[0]));
         }
         else {
+            int iop;
+            npy_bool has_flexible = 0;
+            npy_bool has_object = 0;
+            for (iop = 0; iop < ufunc->nin; iop++) {
+                if (PyArray_ISOBJECT(operands[iop])) {
+                    has_object = 1;
+                }
+                if (PyArray_ISFLEXIBLE(operands[iop])) {
+                    has_flexible = 1;
+                }
+            }
+            if (has_flexible && !has_object) {
+                /*
+                 * DEPRECATED NumPy 1.19, 2020-03 the following check is needed
+                 * to avoid the warning within ResultType. Effectively these
+                 * types could never promote for ufuncs, so disallow them.
+                 * We have to do this before PyArray_ResultType is called,
+                 * since it would give a spurious DeprecationWarning.
+                 */
+                PyErr_Format(PyExc_TypeError,
+                        "No loop matching the specified signature and "
```

Comment:
making this consistent with the other ufuncloop error message may avoid changes on pandas side.

Reply:
Thanks, now that you said that. I actually tried to see what happens to pandas, and fixing this a bit was part of that. Although it seemed like a lot of warnings and I wasn't sure how to turn them to errors quickly in the run to actually see better whether or not they matter.

Fixing tests like `np.array([2, "foo"])` do not matter, they are quick and easy to fix. But if it is relied on somewhere deeper, we may have to see.